### PR TITLE
Add button/menu item for data stream notifications collection page #539

### PR DESCRIPTION
**Pull request by Farmer-Eds-Shed**
Add button/menu item for data stream notifications collection page #539

Adds to settings menu
image

### CODE REVIEW

**Comment by mstenta** [comment]
Looks OK to me! Requesting @paul121's review since he is more familiar with data streams.

One small thought (not a blocker): I could see this being a sub-tab of a new general "Notifications" tab in the future, which may also include other notification type settings (eg: worker task notifications).

FYI I'll plan to rebase and squash these commits before merging.

**Comment by Farmer-Eds-Shed** [comment]
I'll not be offended if it's better suited elsewhere, was bugging me with having nowhere though.
Funny enough I had considered a new notifications tab as I am also working on some notifications modules that could occupy a sub tab.

**Comment by mstenta** [comment]
> Funny enough I had considered a new notifications tab as I am also working on some notifications modules that could occupy a sub tab.

The main question is: what module should be responsible for the "Notifications" tab, if multiple modules need to rely on it. Doesn't make sense to have it in the Data Stream module. And we don't really want it to show up all the time, otherwise it would be empty without any child tabs.

**Comment by Farmer-Eds-Shed** [comment]
Could it not just be it's own module in this case? Not installed by default but required by any new notifications module.

**Comment by paul121** [comment]
Interesting thanks for working on this @Farmer-Eds-Shed !! My first thought is that this seems a little weird in the Settings pages, but it's not awful. Agree that it would look better under a "Notifications" tab...

> The main question is: what module should be responsible for the "Notifications" tab, if multiple modules need to rely on it. Doesn't make sense to have it in the Data Stream module. And we don't really want it to show up all the time, otherwise it would be empty without any child tabs.

Agreed. But perhaps this could just be provided by `farm_settings` ? I do believe that the parent tab won't be displayed if there are no children tabs available or accessible (due to permissions). If it's possible for a "local task" to be provided that doesn't actually link to anything (there wouldn't be a canonical "notifications" page), then `farm_settings` could just provide the notification "Namespace" and let other modules populate it.

Docs for reference: https://www.drupal.org/docs/drupal-apis/menu-api/providing-module-defined-local-tasks#s-define-static-local-tasks

> route_name (required): The machine name of the local task route - this also determines where it's displayed.

Hmm but looks like a `route_name` is required. Not sure if there is a clever way to get around that... maybe we could make a dummy route of some sort? Even if it was simply "Select the tab for the type of notification you would like to configure." I think we would either need a canonical notification page of some sort or dynamic logic to make the Notification tab go to the lowest-weight notification tab that is provided. Hmmm. Any :bulb: ideas @Farmer-Eds-Shed @mstenta ?

_Alternatively...._ at one point I think we had discussed adding it as an Action Link on the data stream view tab. So when you're viewing a data stream, there could be a button "Configure notifications" in the top right. This would bring you to this page to configure *all* notifications, not only notifications for the single data stream, so perhaps that is confusing anyways.

**Comment by Farmer-Eds-Shed** [comment]
> Interesting thanks for working on this

No Problem, It's a feature I want to use myself, so I felt it needs to go somewhere. 
> My first thought is that this seems a little weird in the Settings pages, but it's not awful

Yea, probably can't disagree, it's the least worst of the menu places I experimented with though 😆.

> _Alternatively...._ at one point I think we had discussed adding it as an Action Link on the data stream view tab. So when you're viewing a data stream, there could be a button "Configure notifications" in the top right. This would bring you to this page to configure _all_ notifications, not only notifications for the single data stream, so perhaps that is confusing anyways.

Not sure I like this one either as I agree it could be confusing, it would be better to link to notifications for the single data stream this way, but then it would still make sense for a notifications menu to manage all notifications.

> Hmm but looks like a `route_name` is required. Not sure if there is a clever way to get around that... maybe we could make a dummy route of some sort? Even if it was simply "Select the tab for the type of notification you would like to configure." I think we would either need a canonical notification page of some sort or dynamic logic to make the Notification tab go to the lowest-weight notification tab that is provided. Hmmm. Any 💡 idea

If creating an empty menu is an issue, then may be a basic notifications module is an option, similar to the data streams one but a little less obscure for the average user. I'd imagine the option of notifications for log creation/update could be popular and not weird to be activated alongside any other notification module that could be wanted in the future?  

I still propose using the settings menu for now if implementing the notifications menu may be a bit off,  as somewhere is better than nowhere, I opened the PR as I want to keep in line with farmOS without maintaining minor customizations on my server, but I can just as easily create a module to achieve the same. There are others interested in data streams but are probably unable to find this feature.

**Comment by paul121** [comment]
> I still propose using the settings menu for now if implementing the notifications menu may be a bit off, as somewhere is better than nowhere, I opened the PR as I want to keep in line with farmOS without maintaining minor customizations on my server, but I can just as easily create a module to achieve the same. There are others interested in data streams but are probably unable to find this feature.

Totally agree, thanks for expressing this. Given that this will only appear if that module is enabled, I agree that this is better than nowhere. In the future we could move this without it being a "breaking" change as well.

**Review by paul121 — APPROVED** [review]
I just wanted to review which permission is required to access this `entity.data_stream_notification.collection` route..

I believe it is this `access data_stream_notification overview` which we only grant to `config` roles, so this seems safe to me :+1: 

https://github.com/farmOS/farmOS/blob/fc77c86abe6b142702e7b103cd85e0c207333b79/modules/core/data_stream/modules/notification/data_stream_notification.managed_role_permissions.yml#L3

**Comment by mstenta** [comment]
Thanks for reviewing @paul121! And all the good thoughts @Farmer-Eds-Shed!

>  may be a basic notifications module is an option

I started an initial branch for this on yesterday's dev call: https://github.com/mstenta/farmOS/tree/2.x-notification

We discussed that this core `farm_notification` module could eventually become the home for general notification logic (perhaps extracted/generalized from the `data_stream_notification` module). For now it just provides a "Notifications" tab under settings, which the data stream notification module can depend on.

I think I'll merge this, then rebase this PR on top and put the new menu link under that. Then we can look at generalizing the data stream notification code as a future step.

**Comment by mstenta** [comment]
Updated screenshot:

Screenshot from 2022-08-12 09-21-51

File: modules/core/notification/farm_notification.routing.yml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+farm_notification.settings:
+  path: '/farm/settings/notification'
+  defaults:
+    _form: '\Drupal\farm_notification\Form\NotificationSettingsForm'
+    _title: 'Notification settings'
+  requirements:
+    _permission: 'administer farm notification'
```
Comment (paul121) [inline modules/core/notification/farm_notification.routing.yml:7]: I believe we need to add this to `managed_role_permissions` like here: https://github.com/farmOS/farmOS/blob/fc77c86abe6b142702e7b103cd85e0c207333b79/modules/core/data_stream/modules/notification/data_stream_notification.managed_role_permissions.yml#L3

File: modules/core/data_stream/modules/notification/data_stream_notification.info.yml [hunk 2]
```diff
@@ -5,3 +5,4 @@ package: farmOS
 core_version_requirement: ^9
 dependencies:
   - farm:data_stream
+  - farm:farm_notification
```
Comment (paul121) [inline modules/core/data_stream/modules/notification/data_stream_notification.info.yml:8]: I think we need an update hook to enable `farm_notification` on sites that already have `data_stream_notification` ?

File: modules/core/notification/src/Form/NotificationSettingsForm.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildForm(array $form, FormStateinterface $form_state) {
+
+    // Add a placeholder message.
+    $form['placeholder'] = [
+      '#type' => 'markup',
+      '#markup' => $this->t('Notification modules can add additional tabs to this page.'),
```
Comment (paul121) [inline modules/core/notification/src/Form/NotificationSettingsForm.php]: Could we add a prompt for the user? On mobile or small screens the tab options won't be visible. eg: "Select the tab for the type of notification to configure. Notification modules can add additional tabs to this page."

**Review by paul121 — CHANGES_REQUESTED** [review]

**Comment by mstenta** [comment]
Thanks @paul121! Good eyes...

I made all the changes you requested in two "fixup" commits (to be squashed into the appropriate original commits before merging).

I also made one other small change: I made "Notifications" plural in the tab titles.

Requesting re-review before merge.

**Comment by mstenta** [comment]
PS: I squashed the commits as described and pushed to a new `2.x-notification` branch if you want to see the final result: https://github.com/farmOS/farmOS/compare/2.x...mstenta:farmOS:2.x-notification

**Review by paul121 — APPROVED** [review]
LGTM!

**Comment by mstenta** [comment]
Merged - thanks again @Farmer-Eds-Shed and @paul121!

**Pull request merged**